### PR TITLE
Better testing of reaction to raterule

### LINKED ISSUE #183: Correct converter to raterules for when reactions have only products
see  BIOMD0000001042

### PR DESCRIPTION
**Pull request by skeating**
## Description

Put in a lot more test for the reaction converter
## Motivation and Context

Fixes #183 
also deals with all the edge cases
## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Change in documentation

## Checklist:

- [ ] I have updated all documentation necessary.
- [ ] I have checked spelling in (new) comments.

## Testing
- [X] Testing is done automatically and codecov shows test coverage
- [ ] This cannot be tested automatically

### CODE REVIEW

**Comment by skeating** [comment]
it worked on my machine I'll look at it later

**Comment by fbergmann** [comment]
test failures seem due to c++ namespaces and strict includes as far as i can tell.

**Comment by skeating** [comment]
>## Description
> Put in a lot more test for the reaction converter
> 
> ## Motivation and Context
> Fixes #183 also deals with all the edge cases
> 
> ## Types of changes
> * [ ]  Bug fix (non-breaking change which fixes an issue)
> * [ x] New feature (non-breaking change which adds functionality)
> * [ ]  Breaking change (fix or feature that would cause existing functionality to change)
> * [ ]  Change in documentation
> 
> ## Checklist:
> * [ ]  I have updated all documentation necessary.
> * [ ]  I have checked spelling in (new) comments.
> 
> ## Testing
> * [x]  Testing is done automatically and codecov shows test coverage
> * [ ]  This cannot be tested automatically

File: src/sbml/conversion/test/TestSBMLRateRuleConverter.cpp [hunk 1]
```diff
@@ -727,18 +727,18 @@ create_suite_TestSBMLRateRuleConverter (void)
...
-  tcase_add_test(tcase, test_conversion_raterule_converter_hidden_variable);
-  tcase_add_test(tcase, test_crash_converter);
-  tcase_add_test(tcase, test_model);
-  tcase_add_test(tcase, test_model1);
-  tcase_add_test(tcase, test_model2);
-  tcase_add_test(tcase, test_model3);
-  tcase_add_test(tcase, test_model4);
-  tcase_add_test(tcase, test_model5);
-  tcase_add_test(tcase, test_model6);
+  //tcase_add_test(tcase, test_conversion_raterule_converter_invalid);
```
Comment (fbergmann) [inline src/sbml/conversion/test/TestSBMLRateRuleConverter.cpp]: @skeating can we include these tests again?

File: src/sbml/conversion/test/TestSBMLReactionConverter.cpp [hunk 2]
```diff
@@ -0,0 +1,354 @@
...
+Suite*
+create_suite_TestSBMLReactionConverter(void)
+{
+	bool testing_individual = false;
+	Suite* suite = suite_create("SBMLReactionConverter");
+	TCase* tcase = tcase_create("SBMLReactionConverter");
+	tcase_add_checked_fixture(tcase, Reaction_setup,
+		Reaction_teardown);
+
+	if (testing_individual)
```
Comment (fbergmann) [inline src/sbml/conversion/test/TestSBMLReactionConverter.cpp]: @skeating this seems for testing only, can this be removed?

File: src/sbml/conversion/ExpressionAnalyser.cpp [hunk 3]
```diff
@@ -97,18 +112,15 @@ ExpressionAnalyser::clone() const
...
+  SBMLTransforms::clearComponentValues(mModel);
+  if (mExpressions.size() > 0)
   {
-    if (it->second != NULL)
-    {
-      delete it->second;
-      it->second = NULL;
-    }
+    mExpressions.clear();
   }
```
[the remark this answers is not in the document]
Comment (fbergmann) [reply src/sbml/conversion/ExpressionAnalyser.cpp]: @skeating is that a thing, should we run it through valgrind?

File: src/sbml/conversion/ExpressionAnalyser.cpp [hunk 4]
```diff
@@ -441,41 +453,102 @@ ExpressionAnalyser::detectHiddenSpecies(List * hiddenSpecies)
...
+                currentParent->replaceChild(index, replacement->deepCopy(), false);
+                replacementMade = true;
+            // intentionally, don't delete replacement as it's now owned by currentParent!
+            }
+        } while (currentParent != NULL);
+    }
+    return replacementMade;
 }
 
+
```
Comment (fbergmann) [inline src/sbml/conversion/ExpressionAnalyser.cpp:508]: thats a lot of new lines :D

**Review by fbergmann — APPROVED** [review]
